src/serve/ouroboros/mod.rs

```diff
@@ -105,3 +83,140 @@ async fn handle_blockfetch(db: RollDB, mut peer: PeerServer) -> Result<(), Error
+                    return Some(point.clone());
+                }
+            }
+        }
+    }
+
+    None
+}
+
+fn db_tip_to_protocol(tip: (u64, Hash<32>)) -> Tip {
+    // TODO: get block height from db
+    Tip(Point::Specific(tip.0, tip.1.to_vec()), 0)
+}
+
+#[instrument(skip_all)]
+async fn handle_chainsync(db: RollDB, mut protocol: chainsync::N2NServer) -> Result<(), Error> {
+    let mut intersect = db.crawl_chain_from(None);
+
+    loop {
+        debug!("waiting for request");
+
+        let req = protocol.recv_while_idle().await.map_err(Error::server)?;
+
+        debug!("new client request");
+
+        let tip = db
+            .find_tip()
+            .map_err(Error::server)?
+            .map(db_tip_to_protocol)
+            .unwrap_or(Tip(Point::Origin, 0));
+
+        debug!(?tip, "fetched tip from db");
+
+        match req {
+            Some(x) => match x {
+                chainsync::ClientRequest::Intersect(points) => {
+                    debug!(?points, "intersect request");
+
+                    if let Some(found) = find_valid_intersection(&db, &points) {
+                        intersect = match found {
+                            Point::Origin => db.crawl_chain_from(None),
+                            Point::Specific(x, _) => db.crawl_chain_from(Some(x)),
+                        };
+
+                        protocol
+                            .send_intersect_found(found, tip)
+                            .await
+                            .map_err(Error::server)?;
+                    } else {
+                        protocol
+                            .send_intersect_not_found(tip)
+                            .await
+                            .map_err(Error::server)?;
+                    }
+                }
+                chainsync::ClientRequest::RequestNext => {
+                    let next = intersect.next();
+
+                    if let Some(next) = next {
+                        let (_, hash) = next.map_err(Error::server)?;
+
+                        let block = db
+                            .get_block(hash)
+                            .map_err(Error::server)?
+                            .expect("block content not found");
+
+                        let block = MultiEraBlock::decode(&block).expect("invalid block cbor");
+
+                        let content = HeaderContent {
+                            variant: 1, // TODO
+                            byron_prefix: None,
+                            cbor: block.header().cbor().to_vec(),
+                        };
+
+                        protocol
+                            .send_roll_forward(content, tip)
+                            .await
+                            .map_err(Error::server)?;
+                    } else {
+                        protocol.send_await_reply().await.map_err(Error::server)?;
+                    }
+                }
+            },
+            None => todo!(),
```

Comment:
`None` here signals a ClientDone from the client, so we can probably exit with a log message saying as much and then `return Ok(())` maybe?